<!DOCTYPE html>
<html>
<head>
    <title>Human Regulatory Mutation Database Help</title>
    <meta name="generator" content="Bluefish 2.2.5">
    <meta name="author" content="peter">
    <meta name="date" content="2015-11-04T09:09:09+0100">
    <meta name="copyright" content="">
    <meta name="keywords" content="">
    <meta name="description" content="">
    <meta name="ROBOTS" content="NOINDEX, NOFOLLOW">
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <meta http-equiv="content-type" content="application/xhtml+xml; charset=UTF-8">
    <meta http-equiv="content-style-type" content="text/css">
    <meta http-equiv="expires" content="0">
</head>
<body>
<h1>Human Regulatory Mutation Database Help</h1>
<p>This application is designed to help enter data for the Human Regulatory Mutation Database (HRMD). Each mutation
    that we curate will be stored in the form of an XML file. There will be a single XML file for each mutation that
    also includes other information as explained below. Note the things to do for homozygous and compound heterozygous
    mutations (explained under <b>variant</b>). The application is designed to perform a number of Q/C checks to ensure
    that the data is consistent. This how-to contains information on (i) Setting up your environment; (ii) Entering data
    for a mutation; and (iii) validating your entries.
</p>
<h2>Setting up the HRMDgui App</h2>
<p>
    When you run the application for the very first time, you will need to download some files that HRMDgui uses for
    Q/C. You can download the data directly from the <b>Setup</b> menu. For the HGMD public data and for the
    genome FASTA files and FASTA index (fai) files, you will need to tell the application where you have stored the
    files on your system. HRMDgui downloads the files <i>hp.obo</i> and the EntrezGene info for for H. sapiens (<i>Homo_sapiens.gene_info.gz</i>)
    into a directory called <b>hrmdgui_data</b> (the location of this directory is hard-coded, and is located within the
    same directory from which you start the application. That is, you should probably always start the application from
    the same GIT hub-cloned directory).
</p>
<p>There are two default settings files, one for cancer and one for Mendelian (germline) mutations. You should see these
    in the GUI and can click on them to set up the GUI to take either Mendelian or Cancer mutations.
    Please store your own settings in a file that you do not necessarily need to check into github. This will include
    things like your biocurator name and any path that you have to the GRCh37 genome sequences.
    Basically, set everything up like you want it to be, and choose the <b>save settings file</b> item from the setup
    menu.
<p>
    There are things things you need to do in order to prepare the app for use (basically, go through the items in the
    Setup menu). Note that for the downloads to work from behind a firewall, you need to set the http_proxy and
    ftp_proxy environment variables for your system accordingly.
</p>
<h4>1) Setup the XML directory</h4>
Go to the <b>setup</b> menu and choose <i>XML directory</i>, and navigate to the directory where the HRMD XML files are
located. This is in a github directory (ask me if you do not know where it is, we have not currently publicized this!).
Choose the directory with the File chooser that will appear. Save all new data here and make sure to add the new XML
files to the corresponding github repository.
<h4>2) Download EntrezGene</h4>
Just click on the corresponding entry of the Setup menu and the file <i>Homo_sapiens.gene_info.gz</i> will be downloaded
to the local directory <b>hrmdgui_data</b>.
<h4>3) Set HGMD path</h4>
If you have a copy of HGMD public data, use this menu entry to tell HRMDgui where the file allmut.txt is, which contains
some useful information that you can visualize. Alternatively, you can use the
public website of HGMD to examine these mutations.
<h4>4) Download the hp.obo file</h4>
Go to the <b>setup</b> menu and choose <i>HPO download</i>.
This needs to be done once (and can be updated as necessary). We download a copy of the <b>hp.obo</b> file so that the
App can autocomplete the HPO terms names (the user should just enter the HPO ID (e.g., HP:0001234).
<h4>5) Genome directory</h4>
<p>Here we tell HRMDgui where we have stored the Genome FASTA files and the corresponding fasta index (fai) file. We
    will use this to check whether the wildtype sequence entered for each mutation matches the corresponding genomic
    position. </p>
<p>We will use a FASTA file for each chromosome of the human genome. For now,
    we will use build 37, i.e., GRCh37, which was also known as hg19. We will assume that there is a fasta index file in
    the same directory that has the exact same same as the fasta file except that it ends with "fai" instead of
    "fa".</p>
<p>If you do not have this data on your computer anyway, then you can do the following to set things up</p>
<ul>
    <li>Down load the file <b>chromFa.tar.gz</b> from UCSC. It is 905 M. Remember to use the file from <b>GRCh37</b>
        (i.e., hg19) and not GRCh38!
        Here is the URL: http://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/.
    </li>
    <li>use $tar xvfz to unpack these files (warning: they will unpack without a surrounding directory!) You will get a
        lot of FASTA files including many for scaffolds (we only need the canonical chromosomes
        for now, but the scaffold files will not hurt either)
    </li>
    <li>We will use samtools to create a fasta index file (fai). Download/build it here: http://www.htslib.org/</li>
    <li>The command to create an index is then (for example.fa) simply $ samtools faindex example.fa</li>
    <li>The index file basically will allow us to quickly navigate to the right part of the genome FASTA file to extract
        the sequence we need to check our mutation data (HRMDgui uses the HTS-JDK library to do this!).
    </li>
</ul>
<p>
    If you have downloaded and unpacked the genome fasta files as above (I recommend to use a directory called hg19 for
    this), then you can use the following shell command to create FAI files for
    each of the FASTA files (note that I leave out the chrUn and random scaffolds, they do not have anything of interest
    for this project).
<pre>
for f in `find . -name '*.fa' | grep -v chrUn | grep -v random`
do
    echo "$f"
    samtools faidx $f
done
</pre>
obviously you will need to adjust the command if samtools is not in your path. This is all we need for the genomic
position Q/C routines.
</p>
<h2>Workflow: How to enter data</h2>
<p>Enter the data from top to bottom in the GUI. In most cases, you will download the PDF file of an article about a
    regulatory mutation and will be consulting it as you enter the data.
<h4>1) PubMed abstract</h4>
<p>
    Go to PubMed and find the article you are curating. Mark the checkbox for the article and click on the <i>Display
    Settings</i> button to
    show the format <i>Summary (text)</i>. After you click on this, you will will a summary version of the abstract that
    looks something like the following:
</p>
<pre>1: Robinson PN. Genomic data sharing for translational research and diagnostics. 
Genome Med. 2014 Sep 26;6(9):78. doi: 10.1186/s13073-014-0078-2. eCollection
2014. PubMed PMID: 25473437; PubMed Central PMCID: PMC4254431.
</pre>
<p>Copy the entire text into the first text field of the app. Check the validity/parse using the <i>Test PubMed
    input</i> button </p>
<h4>2) Genome Build</h4>
For now, please use build 37 (called either "<b>GRCh37</b>" or "<b>hg19</b>"). Later, we will use the liftover utility
of UCSC to add data for build 38.
<h4>3) Chromosome and position</h4>
<p>Consult the article you are reading. I have found it helpful to see if the sequence surrounding the mutation position
    is shown somewhere in the article.
    If this sequence is 20 nucleotides or more, you can use the BLAT tool of UCSC Genome Browser to find the
    corresponding position in the genome. If there are
    only a few bases, sometimes you can use guesswork to narrow things down enough to find the corresponding place in
    the genome. For older articles that
    specify the position of a mutation using Genome Build 36 (called either "GRCH36" or "hg18"), you can use the UCSC
    Liftover utility here:
    <a href="https://genome.ucsc.edu/cgi-bin/hgLiftOver" target="_blank"
       title="https://genome.ucsc.edu/cgi-bin/hgLiftOver">https://genome.ucsc.edu/cgi-bin/hgLiftOver</a>.
    There are some articles that are of such low quality that it is simply not possible to reliably identify the
    chromosomal position of the mutation. In these cases, the article should be
    <b>rejected</b>. It may also be worthwhile to consult dbSNP or ClinVar, since some published mutations are entered
    in these databases.</p>
<p>Note that position should be <b>one-based</b>, and not zero-based.</p>
<h4>4) Ref/Alt</h4>
<p>For single-nucleotide variants, Ref and Alt are simply A,C,G, or T.</p>
<p>
    For deletions and insertions, please use the VCF format. Here is the Webpage with the latest details
    (https://github.com/samtools/hts-specs), but if in doubt please ask Peter.
    Just to give a simple example. Let us pretend we have a ten base-pair reference sequence on chromosome "Z"
<pre>
ACGTAAGTCA
</pre>
Let us imagine that the "T" at position 4 is deleted. This results in the sequence
<pre>
ACGAAGTCA
</pre>
It might seem logical to write simple position=4, ref="T", alt="-". VCF format calls instead for this
<pre>
#CHROM POS ID REF ALT (other stuff)
Z 3 . GT G (other stuff)
</pre>
This means that the dinucleotide at position 3-4 is affected and the variant sequence has only a G. For an insertion of
a C between the T at position 4 and the A at position 5, we write
<pre>
#CHROM POS ID REF ALT (other stuff)
Z 4 . T TC . (other stuff)
</pre>
We will use this convention, which will allow us to check the reference sequence and the position even for deletions,
and should allow us a little more possibilities for Q/C-ing the genomic position etc.</p>


<h4>5) Variant status</h4>
<p>We need to enter information about whether the variant is
    <b>heterozygous</b> or <b>homozygous</b>. Note that if the patient has two different heterozygous mutations (i.e.,
    is <b>compound heterozygous</b>), then we enter the second mutation in the second Variant box. In all other cases,
    we just use the first Variant box. Also, note that in some cases, the publications state (for an autosomal recessive
    disease) that "the second mutation could not be found". Also in this case, do not enter anything into the second
    Variant box.
</p>
<p>
    Note that if the first mutation is regulatory and the second mutation is coding (e.g., missense, nonsense, splicing
    etc), then you should use the category "coding" for the second mutation. Remeber that at least one of the mutations
    should be regulatory in order to "qualify" for this database!
</p>
<p>
    Finally, it is a good idea to use the <b>Mutalyzer</b> to check the nomenclature and locationi of coding mutations.
    The Mutalyzer will provide the surrounding genomic sequence for most mutations, and this can be used to identify the
    genomic position of coding mutations using BLAT. It may also be useful to consult with ClinVar or the public version
    of HGMD about this.
</p>
<h4>6) Variant class</h4>
<p>One of
<ol>
    <li><b>promoter</b>. Note that there are no really good definitions of where the promoter is located. Please put
        anything in the 5UTR in the class 5UTR, even if the effect seems to be on the promoter. Probably anything within
        5-10,000 nucleotudes upstream of the transcription start site can be called promoter, but since we will have the
        numbers, we can do the classification automatically later. For now, I have taken the classification as mentioned
        in the original publications.
    </li>
    <li><b>enhancer</b>. Regulatory region that is farther removed from the transcriptional start site than a promoter
    </li>
    <li><b>5' UTR</b></li>
    <li><b>3' UTR</b></li>
    <li><b>microRNAgene</b>. Here we mean any variation that affects the transcript that encodes for a microRNA (note:
        mutations that affect microRNA binding sites should in general be classified as 3UTR)
    </li>
    <li><b>RNP_RNA</b>. Ribonucleoprotein (RNP) RNA component gene. These include ribosome and snRNP</li>
    <li><b>LINC_RNA</b>. Long intergenic non-coding RNA gene</li>
    <li><b>coding</b>. We only incldue coding mutations if the patient being described was compound heterozygous for a
        coding mutation and a regulatory mutation
    </li>
</ol>
Note that the 5UTR DNA sequences often form part of the actual promoter, and in general it is not possible to know if a
variant affects the promoter function or the 5' UTR function
(which is of course in the mRNA and can affect the stability of the transcript). If a mutation is located in the 5UTR,
then please enter 5UTR even if the effect is on the promoter.
The data base and downstream analysis just has to know about this. In some cases, a mutation may be both 5UTR and
promoter etc. Please enter the category that seems most relevant. We will automatically generate these annotations using
Jannovar anyway, so even variants with multiple categories will be correctly classified.
</p>
<p>
    Note again that the category "coding" should only be used for the "second" mutation in compound heterozygous cases.
    At some point we may want to consider adding other classes, but none of the old data will be affected by a new class
    (e.g., silencer).
</p>
<h4>7) Validation</h4>
<p> We will only include a mutation in the HRMD if there is adequate evidence for its pathogenicity. As a general rule,
    there should be
    some experimental evidence for the mutation changing gene regulation of a target gene in some way. For some heavily
    studied genes, we
    will accept a mutation if it seems to be very similar to other published mutations (e.g., it lies on the same
    predicted transcription
    factor binding site as another mutation for which experimental evidence is available). Add as much evidence as
    possible. It is expected
    that at least one of the evidence categories will apply to each mutation.

</p>
<ul>
    <li><b>Reporter</b>. Luciferase assay (or the similar CAT assay) to judge transcriptional activity. Indicate whether
        the mutation is associated with increased activity (<b>up</b>) or
        decreased activity (<b>down</b>) as compared with the wildtype construct (in percent).
    </li>
    <li><b>EMSA</b>. EMSA (electrophoretic mobility shift assay). This is used to indicate whether a protein binds to a
        given DNA sequence. For our purposes, we are referring to the
        protein affected by the mutation. Enter the corresponding protein if there is a change in binding. Enter the
        Entrez Gene ID and Gene Symbol of the protein that is affected by
        the mutation (usually a transcription factor)
    </li>
    <li><b>Cosegregation</b> Enter yes if the mutation cosegregates with the disease in the family being investigated.
    </li>
    <li><b>Comparability</b> This is the weakest evidence class. Enter yes if the reason for believing that the mutation
        is pathogenic is simply that it is comparable to
        other published regulatory mutations in the gene.
    <li><b>Other</b> This is for any other kind of experimental assay that shows an effect of a regulatory or non-coding
        mutation. Note that for now the categories are hard coded into
        the Java code, this should be put into some kind of configuration file in the future.
        The categories are at present:
        <ol>
            <li><b>Telomerase</b>. Telomerase lengthening assay.</li>
        </ol>
</ul>


<h4>8) Disease data</h4>
Set the database (please use the OMIM id if at all possible). For OMIM, use the <b>phenotype</b> id, and not the gene
id!.
<ol>
    <li>Database: one of OMIM or ORPHANET (use drop-down menu)</li>
    <li>The disease. Please use a lower-case form of the canonical name, i.e., do not include all of the synonyms in
        upper-case letters.
    </li>
    <li>The database ID. For OMIM; this will be a number like 614321</li>
</ol>
<h4>9) HPO terms</h4>
<p>To enter the HPO data, set the number of HPO terms you want and click on the <i>add HPO Terms</i> button.
    Note that if you find you do not have enough, you can add additional terms with this button too.</p>
<p>Enter the HPO id and "tab" to the next field, which should autocomplete with the term name. Check that you have
    gotten the correct id.
</p>
<p>
    I have found it to be helpful to open OBO-Edit with the HPO and to search here for the terms that apply to the
    patient described in the publication
    we are annotating. Then, basically copy and paste the data to the HRMDgui HPO window. Note that this seems to work
    better under linux than
    under Mac, which still does not have the greatest Java support.
</p>
<h4>10) ID</h4>
<p>The ID (patient/family identifier) is a free-text string that represents the ID used to designate the affected
    individual or family in the
    original paper. For instance, <b>family 3</b>. Note that we usually include all of
    the regulatory mutations in a given paper, but if little clinical data is given, and the phenotype is identical for
    two families,
    then it is OK to enter <b>family 3 and family 7</b>, say.
<h4>11) Metadata</h4>
<p>Many of the individual papers about regulatory mutations have a lot of interesting additional information that is
    more or less heterogeneous. We would like to capture the
    most salient points in a free text that will be displayed on the HRMD Website. For instance, here is an example
    Metadata
<pre>The mutation is located in a 400-bp sequence located 25 kb downstream of PTF1A 
(the gene for pancreas-specific transcription factor 1a). This region acts as a 
developmental enhancer of PTF1A and that the mutations abolish enhancer activity.
The mutation was shown to abolish binding of FOXA2 (Supplementary Figure 8 of 
Wheedon et al., 2014).</pre>
</p>
<h4>12) Target Gene</h4>
In presumably almost all cases, we will know the target gene of a regulatory mutation that has been published. We enter
two bits of information:
<ol>
    <li>The Entrez Gene ID</li>
    <li>The gene symbol</li>
</ol>
<p>For instance, for the target gene <b>pancreas specific transcription factor, 1a</b> we would enter the ID 256297 and
    the symbol PTF1A.</p>
<h4>13) Regulator</h4>
<p>This is a free text field that can contain text such as
<ul>
    <li>Promoter of PTF1A</li>
    <li>5 UTR of PTF1A</li>
    <li>3 UTR of PTF1A</li>
    <li>Pancreatic progenitor–specific enhancer</li>
</ul>
There does not currently exist a good standard nomenclature for enhancers and promoters that we could use for this
purpose, so free text seems the best option. In addition, other features of the
Website will be able to infer what the regulatory element is from the position of the mutation.
</p>
<h4>14) Saving your data</h4>
<p>You can use the <i>See/Edit XML</i> button to open an editor where you can check and if necessary correct the XML
    file.</p>
<p>When everything is ready, save the file. If you like, you can save the file directly to the main XML directory
    (remember to check it in with Subversion). The naming scheme would be something like
    <i>Weedon-PTF1A-2013-1.xml, Weedon-PTF1A-2013-2.xml, ...., Weedon-PTF1A-2013-9.xml</i> for the various mutations
    described in the publication

<pre>Weedon MN et al. Recessive mutations in a 
distal PTF1A enhancer cause isolated pancreatic agenesis. Nat Genet. 2014
Jan;46(1):61-4. doi: 10.1038/ng.2826. Epub 2013 Nov 10. PubMed PMID: 24212882;
</pre>
The app will open a file-save dialog in the directory that you specified in the Setup to contain all of the mutation XML
files.
</p>


<h2>Validation and Q/C</h2>
The app has three functions to use to validate the data that has been entered into the database, all available via the
<b>Validate</b> menu.
<h4>1) Validate XML files</h4>
Make sure that the mutation ID for the files you create is unique and valid (A message will be shown if two or more
mutations have been assigned the same mtuation ID).
It will also check that the HPO terms are valid (i.e., that the HPO id, such as HP:0001234, matches the term name given
in the hp.obo file. This will remind us to update the name in case it is changed etc.
or if something else happens to the HPO term -- merge etc).
<h4>2) Validate Entrez Gene</h4>
This checks whether the Entrez ID you entered matches the gene symbol you entered. It also compares the chromosome you
entered with the chromosome name indicated in the
<i>Homo_sapiens.gene_info.gz</i> file.
<h4>3) Validate genomic positions</h4>
<p>Click on the Baseball and the App will check that the reference sequence and position as entered actually match the
    corresponding sequences on the genome FASTA files,
    and show a useful message. If there are mistakes, probably just copy the message to a text editor and start to
    examine the problem.</p>


<h2>Cancer</h2>
<p>We are beginning a new project to biocurate cancer mutations. For now, the main difference is that HPO terms are not
    required, and you should use disease names from the
    NCI thesaurus (https://ncit.nci.nih.gov/ncitbrowser/) to describe the type of cancer you are biocurating.</p>
</body>
</html>
